x/ecocredit/client/marketplace/tx.go

```diff
@@ -0,0 +1,133 @@
+		Short: "Buy ecocredits from a specific sell order",
+		Long: "Purchase ecocredits from a specific sell order. DisableAutoRetire can be set to false to retire the credits immediately upon purchase." +
+			"When set to true, credits will be received in a tradable state, IF AND ONLY IF the sell order also has auto retire disabled. " +
+			"NOTE: The bid price is the price paid PER credit. The total cost will be quantity * bid_price.",
+		Example: "regen tx ecocredit buy-direct 194 300 40regen true --retirement-location=US-NY",
+		Args:    cobra.ExactArgs(4),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientTxContext(cmd)
+			if err != nil {
+				return err
+			}
+			sellOrderIdStr, qtyStr, bidPriceStr, autoRetireStr := args[0], args[1], args[2], args[3]
+
+			sellOrderId, err := strconv.ParseUint(sellOrderIdStr, 10, 64)
+			if err != nil {
+				return err
+			}
+
+			bidPrice, err := sdk.ParseCoinNormalized(bidPriceStr)
+			if err != nil {
+				return err
+			}
+
+			disableAutoRetire, err := strconv.ParseBool(autoRetireStr)
+			if err != nil {
+				return err
+			}
+			var retireLocation string
+			retireLocation, err = cmd.Flags().GetString(FlagRetirementLocation)
+			if err != nil {
+				return err
+			}
+
+			msg := marketplace.MsgBuyDirect{
+				Buyer: clientCtx.GetFromAddress().String(),
+				Orders: []*marketplace.MsgBuyDirect_Order{
+					{
+						SellOrderId:        sellOrderId,
+						Quantity:           qtyStr,
+						BidPrice:           &bidPrice,
+						DisableAutoRetire:  disableAutoRetire,
+						RetirementLocation: retireLocation,
+					},
+				},
+			}
+			if err := msg.ValidateBasic(); err != nil {
+				return err
+			}
+
+			return tx.GenerateOrBroadcastTxCLI(clientCtx, cmd.Flags(), &msg)
+		},
+	}
+	cmd.Flags().String(FlagRetirementLocation, "", "the location to use for retirement when auto retire is true.")
+	flags.AddTxFlagsToCmd(cmd)
+	return cmd
+}
+
+// TxBuyDirectBatch returns a transaction command for a batch direct buy order using a json file.
+func TxBuyDirectBatch() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "buy-direct-batch [name_of_file.json]",
+		Short: "Buy ecocredits from multiple sell orders",
+		Long: "Batch purchase ecocredits using a json file. DisableAutoRetire can be set to false to " +
+			"retire the credits immediately upon purchase. When set to true, credits will be received in a tradable state, " +
+			"IF AND ONLY IF the sell order also has auto retire disabled. NOTE: The bid price is the price paid PER credit. " +
+			"The total cost will be quantity * bid_price.",
+		Example: strings.TrimSpace(`regen tx ecocredit buy-direct-batch batch.json
+		where batch.json has the following form:
+		[
+			{
+			   "sell_order_id": 52,
+			   "quantity": "32.5",
+			   "bid_price": {"denom": "uregen", "amount": "32000000"},
+			   "disable_auto_retire": false,
+			   "retirement_location": "US-NY"
+			},
+		]`),
```

Comment:
We should probably be consistent with how we do batch orders in the create sell order and update sell order commands. We use yaml there but we use json here.

Any reason you chose json over yaml? Should we update here or there? cc @aleem1314

Reply:
ill look into adding tests as well

Reply:
so looks like our CLI integration tests are still commented out until our test suite is wired up to work with ORM. i think we had this issue when we did our v3 sprint. 

should be able to add tests for this once the wiring mayhem is done

Reply:
> until our test suite is wired up to work with ORM

This should be good to go now that https://github.com/regen-network/regen-ledger/pull/901 is merged, correct?

Reply:
not exactly,, https://github.com/regen-network/regen-ledger/pull/901/files#diff-d2a466a8c829cba98b8d562a0fc4193003da21c70943e4feb694c48142451612R52-R61

#901 switched the client impls to use the updated types, but the tests are not updated yet.